jax/interpreters/masking.py

```diff
@@ -432,6 +452,46 @@ def todo(vals):
       return map(partial(MaskTracer, trace), vals, shapes)
     return vals, todo
 
+  def process_map(self, map_primitive, f, tracers, params):
+    vals, shapes = unzip2((t.val, t.polymorphic_shape) for t in tracers)
+    if not any(is_polymorphic(s) for s in shapes):
+      return map_primitive.bind(f, *vals, **params)
+    mapped_invars = params['mapped_invars']
+    logical_env, padded_env = shape_envs
+    poly_axis_size, = {shape[0]
+                       for shape, mapped_invar in zip(shapes, mapped_invars)
+                       if mapped_invar}
+    padded_axis_size = eval_poly(poly_axis_size, padded_env)
+    env_keys, padded_env_vals = unzip2(sorted(padded_env.items()))
+    params = dict(
+        params,
+        mapped_invars=((False,) * len(logical_env) + params['mapped_invars']),
+        axis_size=padded_axis_size)
+    shapes = tuple(shape[1:] if mapped_invar else shape
+                   for shape, mapped_invar in zip(shapes, mapped_invars))
+    logical_env_vals = tuple(logical_env[k] for k in env_keys)
+    padded_env = (env_keys, padded_env_vals)
+    # TODO(j-towns): if poly_axis_size happens to be a constant then just use
+    # mask_subtrace here
+    f, shapes_out = mapped_mask_subtrace(f, self.master, params['axis_name'],
+                                         poly_axis_size, shapes, padded_env)
+    if 'donated_invars' in params:
```

Comment:
We can punt this to later, but I'd rather factor this like `pe.call_param_updaters` / `ad.call_param_updaters`.